src/Plugins/ComplexCore/src/ComplexCore/Filters/Algorithms/RawBinaryReader.cpp

```
@@ -0,0 +1,245 @@
+  std::byte* chunkptr = reinterpret_cast<std::byte*>(dataArray->template getIDataStoreAs<DataStore<T>>()->data());
+
+  // Now start reading the data in chunks if needed.
+  size_t chunkSize = std::min(numBytesToRead, k_DefaultBlocksize);
+
+  size_t master_counter = 0;
+  while(master_counter < numBytesToRead)
+  {
+    size_t bytes_read = std::fread(chunkptr, sizeof(std::byte), chunkSize, f);
+    chunkptr += bytes_read;
+    master_counter += bytes_read;
+
+    size_t bytesLeft = numBytesToRead - master_counter;
+
+    if(bytesLeft < chunkSize)
+    {
+      chunkSize = bytesLeft;
+    }
+  }
+
+  if(endian == k_EndianCheck)
+  {
+    dataArray->byteSwapElements();
+  }
+
+  std::fclose(f);
+
+  return k_RbrNoError;
+}
+} // namespace
+
+using namespace complex;
+
+RawBinaryReader::RawBinaryReader(DataStructure& dataStructure, RawBinaryReaderInputValues* inputValues, const IFilter* filter, const IFilter::MessageHandler& mesgHandler)
+: m_DataStructure(dataStructure)
+, m_InputValues(inputValues)
+, m_Filter(filter)
+, m_MessageHandler(mesgHandler)
+{
+}
+
+RawBinaryReader::~RawBinaryReader() noexcept = default;
+
+Result<> RawBinaryReader::operator()()
+{
+  return execute();
+}
+
+// -----------------------------------------------------------------------------
+Result<> RawBinaryReader::execute()
+{
+  IDataArray* binaryIDataArray = m_DataStructure.getDataAs<IDataArray>(m_InputValues->createdAttributeArrayPathValue);
+  if(binaryIDataArray == nullptr)
+  {
+    return MakeErrorResult(k_RbrDAError, "Can't obtain DataArray from path '" + m_InputValues->createdAttributeArrayPathValue.toString() + "'.");
+  }
+
+  if(binaryIDataArray->getNumberOfComponents() != static_cast<size_t>(m_InputValues->numberOfComponentsValue))
+  {
+    return MakeErrorResult(k_RbrComponentError, "Failed to acquire DataArray from path '" + m_InputValues->createdAttributeArrayPathValue.toString() + "' with the correct number of components.");
+  }
```

In general, I'd say we should throw an exception if something was already validated in preflight instead of returning an error as something more fundamental will have gone wrong.